Added support for loading Python at runtime (supports both Python 3 and 2)

pvieito opened the pull request:
Added new PythonLibrary struct to load Python library and symbols dynamically at runtime

- This adds support for both Python 3 and 2 with the same codebase.
- You can force a version of Python with the PYTHON_VERSION environment variable, a Python library path with PYTHON_LIBRARY and enable logging with PYTHON_LOADER_LOGGING.
- The implementation comes from the PythonKit Swift package (https://github.com/pvieito/PythonKit) removing the Foundation dependency.

pvieito committed:
Added new PythonLibrary struct to load Python library and symbols dynamically at runtime

- This adds support for both Python 3 and 2 with the same codebase.
- You can force a version of Python with the PYTHON_VERSION environment variable, a Python library path with PYTHON_LIBRARY and enable logging with PYTHON_LOADER_LOGGING.
- The implementation comes from the PythonKit Swift package (https://github.com/pvieito/PythonKit) removing the Foundation dependency.

xwu commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+    }
+    
+    for majorVersion in supportedMajorVersions {
+      for minorVersion in supportedMinorVersions {
```
Would this be more elegant if there were simply an array of tuples called `supportedVersions: [(major: Int, minor: Int?)]`?

Such a tuple could also be manually populated (i.e., `[(3, nil), (3, 8), (3, 7), /* ... */]`). There's no need for supporting nonexistent Python versions (indeed, that would be an oxymoron).

pvieito replied:
Removing support for "nonexistent" Python versions means that old Swift distributions (or even any distributed executable that statically linked the Python module) would not work with any 3.x Python version released after it, even though the Python ABI is marked as stable for all the Python 3 lifetime. Also, it would require maintaining the list and deciding when it's time to add new versions.

Take in account that while in Windows the Python distribution includes both `python3.dll` and `python37.dll` in macOS and Linux we have to specify the exact minor version to load it (`Python.framework/Versions/3.7/Python` or `libpython3.7m.so`).

xwu replied:
If the ABI is stable for all versions of Python 3, then is there a need for a list of supported minor versions?

pvieito replied:
@xwu Read the second paragraph.

xwu replied:
I'm afraid I don't understand. Why do we need a list of supported minor versions?

xwu replied:
I understand that. What I don't understand is why we need a static array of "supported" minor versions if support is not contingent on the minor version.

xwu replied:
We can work on this in a follow-up.

There will never be a version of Python 2 after 2.7 for instance, and it makes no sense to look for one; if "Python 2.30" existed, we certainly wouldn't want to support it. Likewise, is Python 2.0 really supported? This is all to say that iterating over the same minor versions for every major version is not ideal.

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Methods of `PythonLibrary` required to load the Python library.
+extension PythonLibrary {
```
Make the extension private and drop ‘private’ from members.

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+    return pythonLibraryHandle
+  }
+  
+  private static func getPythonLibraryHandle() -> UnsafeMutableRawPointer? {
```
Avoid using “get” or “set” in function names.

Plus, I think this can be a computed property.

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Methods of `PythonLibrary` used for logging messages.
+extension PythonLibrary {
```
Is this used in other files? If not, make the extension private.

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Methods of `PythonLibrary` required to read the environment variables.
+extension PythonLibrary {
```
Should this be private too?

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+
+// Methods of `PythonLibrary` used for logging messages.
+extension PythonLibrary {
+  static func log(message: String) {
```
Argument label “message” doesn’t add any clarity. Why not just log(_:)?

rxwei commented on stdlib/public/Python/PythonLibrary.swift:
```diff
@@ -0,0 +1,184 @@
+  }
+  
+  static func loadSymbol<T>(
+    name: String, legacyName: String? = nil, signature: T.Type
```
“signature” is an overloaded word. I’d suggest using “type:”.

pvieito committed:
Style and structure changes

rxwei reviewed (approved):
LGTM. Thanks!

We will also need an API that enables the user to select a Python version. Something like: `Python.useVersion(3, 6)`. It can be added in a separate patch.

pvieito commented:
The function should be on `PythonLibrary` as merely using `Python` already calls `Py_Initialize()`.

Something like this would do the trick:

```swift
// Methods of `PythonLibrary` required to set the Python version.
extension PythonLibrary {
  private static func useVersion(_ version: PythonVersion) {
    PythonLibrary.Environment.version.set(version.versionString)
  }
  
  public static func useVersion(_ major: Int, _ minor: Int?) {
    let version = PythonVersion(major: major, minor: minor)
    self.useVersion(version)
  }
}
```

Example:

```swift
import Python

PythonLibrary.useVersion(2) // Should be called before any references to `Python`
let sys = try Python.attemptImport("sys")
print(sys.version)
```

pvieito committed:
Adding support for setting a Python version on-process

rxwei commented:
IMO, using a `PythonVersion` type is unnecessary complexity and is not intuitive. I don't think we need any extra types to represent the python version, or even separate major from minor from an API perspective. The dlopen-based API is fully dynamic, extra types won't give us more safety than just using a `Float`. How about just starting with a very simple API:

```swift
public extension PythonLibrary {
  static func useVersion(_ version: Float) {
    ... // do version tuple conversion here and set environment
  }
}
```

pvieito commented:
Take in account that the `PythonVersion` struct is private and it is also used on `loadPythonLibrary(at:version:)`. Also, `Float` is a really bad idea: `3.10` vs `3.1`.

pvieito committed:
Default minor version to nil

rxwei commented:
> Also, `Float` is a really bad idea: `3.10` vs `3.1`.

Ok, I'm sold.

> The public method will be: `static func useVersion(_ major: Int, _ minor: Int? = nil)`.

Adding a single method for now and making `PythonVersion` private sounds good to me.

pvieito committed:
Cleaned private method

pvieito commented:
> Adding a single method for now and making `PythonVersion` private sounds good to me.

Great, done! 👍🏻

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
CI is dead. Gotta wait a while.

rxwei commented:
Could you add new files to `CMakeLists.txt`?

pvieito commented:
@rxwei Should I also remove the `CPython` module?

pvieito committed:
Update CMakeLists adding new files and removing CPython module

pvieito commented:
@swift-ci please test tensorflow

dan-zheng commented:
@pvieito Yes, please remove `CPython`. I see you've already done so. :)

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

pvieito committed:
Force setting an non-nil value type for Linux

dan-zheng commented:
I believe the `unknown expression syntax exists in the source` errors are related to libSyntax verification, which is triggered by the flags `-Xfrontend -verify-syntax-tree`.

Could you try building PythonKit on Linux with those extra flags? Unfortunately, this seems like a libSyntax deficiency. [`TypeNodes.py`](https://github.com/apple/swift/blob/tensorflow/utils/gyb_syntax_support/TypeNodes.py) may not yet have support for function conventions.

@nkcsgexi: could you confirm whether libSyntax supports function conventions like `@convention(c)`?

pvieito commented:
@dan-zheng Exactly:

```
PythonKit$ swift build -Xswiftc -Xfrontend -Xswiftc -verify-syntax-tree   
Compile Swift Module 'PythonKit' (4 sources)
Compile Swift Module 'LoggerKit' (12 sources)
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:39:10: error: unknown expression syntax exists in the source
  type: (@convention(c) () -> ()).self
         ^
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:44:10: error: unknown expression syntax exists in the source
  type: (@convention(c) (PyObjectPointer?) -> ()).self
         ^
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:49:10: error: unknown expression syntax exists in the source
  type: (@convention(c) (PyObjectPointer?) -> ()).self
         ^
...
```

dan-zheng commented:
@pvieito Hmm, that's unfortunate. Unless we can disable `-verify-syntax-tree` for the Python module, this PR is blocked by libSyntax support for this usage of `@convention`.

@nkcsgexi: Could you please provide some advice on how to add libSyntax support for this usage of `@convention`? (Basically, how to resolve [SR-9316](https://bugs.swift.org/browse/SR-9316)). That would be greatly appreciated!

rxwei commented:
The libSyntax issue is a bit surprising. There's no instance of getting the metatype of a `@convention(c)` function in stdlib. But let's not get blocked on this. You can work around this by first making `type` a default argument:

```swift
static func loadSymbol<T>(
  name: String, legacyName: String? = nil, type: T.Type = T.self
) -> T
```

Then change your call sites to provide a contextual type.

```swift
let PyList_New: @convention(c) (Int) -> PyObjectPointer? =
  PythonLibrary.loadSymbol(name: "PyList_New")
```

pvieito committed:
Redefined Python symbols to avoid libSyntax bug (SR-9316)

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please clean test tensorflow

rxwei commented:
It's still failing possibly because of incorrectly configured cmake. You mentioned you couldn't test it on your machine, is that right? If so, I can take it from here.

rxwei committed:
Update tests

rxwei commented:
@pvieito I'm currently debugging some failing tests:

```console
[ RUN      ] PythonRuntime.CheckVersion
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 14
stdout>>> expected: "2.7." (of type Swift.String)
stdout>>> actual: "3.7." (of type Swift.Substring)
[     FAIL ] PythonRuntime.CheckVersion
[ RUN      ] PythonRuntime.PythonList
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 22
stdout>>> expected: <type 'list'> (of type Python.PythonObject)
stdout>>> actual: <class 'list'> (of type Python.PythonObject)
[     FAIL ] PythonRuntime.PythonList
[ RUN      ] PythonRuntime.PythonDict
[       OK ] PythonRuntime.PythonDict
[ RUN      ] PythonRuntime.Range
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.Range
[ RUN      ] PythonRuntime.PartialRangeFrom
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PartialRangeFrom
[ RUN      ] PythonRuntime.PartialRangeUpTo
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PartialRangeUpTo
[ RUN      ] PythonRuntime.BinaryOps
[       OK ] PythonRuntime.BinaryOps
[ RUN      ] PythonRuntime.Comparable
stderr>>> Fatal error: 'try!' expression unexpectedly raised an error: Python exception: '<' not supported between instances of 'int' and 'str': file /Users/rxwei/Development/Swift/swift-tf-source/swift/stdlib/public/Python/Python.swift, line 1173
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] PythonRuntime.Comparable
[ RUN      ] PythonRuntime.Hashable
[       OK ] PythonRuntime.Hashable
[ RUN      ] PythonRuntime.RangeIteration
[       OK ] PythonRuntime.RangeIteration
[ RUN      ] PythonRuntime.Errors
[       OK ] PythonRuntime.Errors
[ RUN      ] PythonRuntime.Tuple
[       OK ] PythonRuntime.Tuple
[ RUN      ] PythonRuntime.MethodCalling
[       OK ] PythonRuntime.MethodCalling
[ RUN      ] PythonRuntime.ConvertibleFromPython
[       OK ] PythonRuntime.ConvertibleFromPython
[ RUN      ] PythonRuntime.PythonConvertible
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PythonConvertible
PythonRuntime: Some tests failed, aborting
```

It seems to be loading Python 3 by default. However, even if I explicitly call `useVersion(2, 7)`, the `CheckVersion` test still produces 3.7.

```swift
PythonLibrary.useVersion(2, 7)

PythonRuntimeTestSuite.test("CheckVersion") {
  let sysModule = Python.import("sys")
  let version = String(sysModule.version)!
  expectEqual("2.7.", version.prefix(4))
}
```
```console
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 14
stdout>>> expected: "2.7." (of type Swift.String)
stdout>>> actual: "3.7." (of type Swift.Substring)
```

pvieito commented:
It seems that we have to reference the symbol:

```swift
guard isType(pythonObject, type: PySlice_Type) else { return nil }
```

Not its address to it:

```swift 
guard isType(pythonObject, type: &PySlice_Type) else { return nil } // Crashes
```

rxwei committed:
`init(owning:)` should retain, while `init(borrowing:)` should not.

rxwei committed:
Use loaded symbol values instead of their addresses as the type argument to `isType(_:type:)`.

rxwei commented:
> I suppose we should modify the symbol definition (to match the real Python C header) instead the calling site?

Symbol definitions are correct. It's just that we don't need these type variables to be `var`s anymore since they are addresses to the actual objects.

pvieito commented:
Ok, great!

rxwei committed:
Make loaded symbols `let`s since they are themselves the address we need

rxwei committed:
Minor clean-up.

- `()` -> `Void`.

- Drop parentheses at the result position from single-result function signatures.

- Replace !(==) with != in `isType(_:type:)`.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

pvieito commented:
@rxwei Please, remove also the symbol `_Py_TrueStructb` (seems to be a copy-paste typo).

rxwei committed:
Delete `_Py_TrueStructb`

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux GPU

rxwei commented:
@swift-ci please test tensorflow Linux GPU

rxwei committed:
Fix ParsableInterface tests

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS CPU

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow Linux GPU

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

rxwei commented:
@swift-ci please test tensorflow macOS

H-World commented:
@pvieito  我想知道在Mac上怎么使用指定版本的python，现在我使用PythonLibrary.useVersion(3)，会崩溃Python library not found. Set the PYTHON_LIBRARY environment variable with the path to a Python library.但是我又找不到方法去指定PYTHON_LIBRARY
